### ISSUE TITLE
Enabling both IPv4 and IPv6 breaks socket binding, Jellyfin won't start

### ISSUE THREAD

**Issue by mooninite** [jellyfin/jellyfin#5340]
**Describe the bug**

**System (please complete the following information):**
 - OS: Fedora 33 x86_64
 - Virtualization: None
 - Clients: Browsers only
 - Browser: Firefox 86, Chrome
 - Jellyfin Version: 10.7.0 RC4
 - Playback: N/A
 - Installed Plugins: Playback reporting status, TVDB
 - Reverse Proxy: apache
 - Base URL: none
 - Networking: Host
 - Storage: local

**To Reproduce**

1. Enable "Enable IPv4" and "Enable IPv6" in Networking web UI.
2. Bind to local address is empty.
3. Restart Jellyfin

**Expected behavior**

Jellyfin listening on any IP (v4 or v6) on port 8096.

**Logs**

<pre>
[2021-03-01 14:26:16.304 -06:00] [INF] [1] Main: Kestrel listening on "Any IP4 Address"
[2021-03-01 14:26:16.305 -06:00] [INF] [1] Main: Kestrel listening on "Any IP6 Address"
[2021-03-01 14:26:16.510 -06:00] [FTL] [1] Microsoft.AspNetCore.Server.Kestrel: Unable to start Kestrel.
System.IO.IOException: Failed to bind to address http://[::]:8096: address already in use.
 ---> Microsoft.AspNetCore.Connections.AddressInUseException: Address already in use
 ---> System.Net.Sockets.SocketException (98): Address already in use
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.<Bind>g__BindSocket|13_0(<>c__DisplayClass13_0& )
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.<Bind>g__BindSocket|13_0(<>c__DisplayClass13_0& )
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.Bind()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportFactory.BindAsync(EndPoint endpoint, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.TransportManager.BindAsync(EndPoint endPoint, ConnectionDelegate connectionDelegate, EndpointConfig endpointConfig)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.<>c__DisplayClass29_0`1.<<StartAsync>g__OnBind|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.EndpointsStrategy.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(IEnumerable`1 listenOptions, AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
[2021-03-01 14:26:16.515 -06:00] [ERR] [1] Main: Kestrel failed to start! This is most likely due to an invalid address or port bind - correct your bind configuration in system.xml and try again.
[2021-03-01 14:26:16.516 -06:00] [FTL] [1] Main: Error while starting server.
System.IO.IOException: Failed to bind to address http://[::]:8096: address already in use.
 ---> Microsoft.AspNetCore.Connections.AddressInUseException: Address already in use
 ---> System.Net.Sockets.SocketException (98): Address already in use
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.<Bind>g__BindSocket|13_0(<>c__DisplayClass13_0& )
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.<Bind>g__BindSocket|13_0(<>c__DisplayClass13_0& )
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketConnectionListener.Bind()
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.SocketTransportFactory.BindAsync(EndPoint endpoint, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.TransportManager.BindAsync(EndPoint endPoint, ConnectionDelegate connectionDelegate, EndpointConfig endpointConfig)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.<>c__DisplayClass29_0`1.<<StartAsync>g__OnBind|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindEndpointAsync(ListenOptions endpoint, AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.ListenOptions.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.EndpointsStrategy.BindAsync(AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.AddressBinder.BindAsync(IEnumerable`1 listenOptions, AddressBindContext context)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.WebHost.StartAsync(CancellationToken cancellationToken)
   at Jellyfin.Server.Program.StartApp(StartupOptions options)
</pre>

**Screenshots**

**Additional context**

Here is the C# code to listen to both ipv4 and ipv6 addresses on the same port:
https://web.archive.org/web/20130909044601/http://blogs.msdn.com/b/malarch/archive/2005/11/18/494769.aspx

**Comment by BaronGreenback** [comment]
Could you upload your network.xml please?

**Comment by mooninite** [comment]
My default network.xml is
<pre>
  <EnableIPV6>false</EnableIPV6>
  <EnableIPV4>true</EnableIPV4>
</pre>
I upgraded my 10.6.4 instance to 10.7.0 RC4 and was watching the server log for any messages to be considered about. I saw messages about IPv6:
<pre>
[WRN] [17] Jellyfin.Networking.Manager.NetworkManager: IPv6 is disabled in Jellyfin, but enabled in the OS. This may affect how the interface is selected.
</pre>
I wanted to "solve" it by enabling IPv6 and that is when I encountered the socket binding issue. Disabling IPv6 in network.xml allows Jellyfin to start.

Nothing besides Jellyfin listens on 8096. 
[network.txt](https://github.com/jellyfin/jellyfin/files/6064520/network.txt)

**Comment by aussetg** [comment]
Can confirm the same problem.

**Comment by BaronGreenback** [comment]
Can you upload the top part of the log that looks like

```
[2021-03-01 15:50:58.495 +00:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Defined LAN addresses : "[127.0.0.1/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16]"
[2021-03-01 15:50:58.500 +00:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Defined LAN exclusions : "[]"
[2021-03-01 15:50:58.502 +00:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using LAN addresses: "[127.0.0.1/8,10.0.0.0/8,172.16.0.0/12,192.168.0.0/16]"
[2021-03-01 15:50:58.511 +00:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using bind addresses: "[127.0.0.1/32]"
[2021-03-01 15:50:58.513 +00:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using bind exclusions: "[]"
```

**Comment by mooninite** [comment]
<pre>
[2021-03-01 14:26:14.001 -06:00] [INF] [1] Main: Jellyfin version: "10.7.0"
[2021-03-01 14:26:14.014 -06:00] [INF] [1] Main: Environment Variables: ["[JELLYFIN_CACHE_DIR, /var/cache/jellyfin]", "[JELLYFIN_WEB_OPT, --webdir=/usr/share/jellyfin-web]", "[JELLYFIN_DATA_DIR, /var/lib/jellyfin]", "[JELLYFIN_CONFIG_DIR, /etc/jellyfin]", "[JELLYFIN_RESTART_OPT, --restartpath=/usr/libexec/jellyfin/restart.sh]", "[JELLYFIN_LOG_DIR, /var/log/jellyfin]"]
[2021-03-01 14:26:14.016 -06:00] [INF] [1] Main: Arguments: ["/usr/lib64/jellyfin/jellyfin.dll", "--webdir=/usr/share/jellyfin-web", "--restartpath=/usr/libexec/jellyfin/restart.sh"]
[2021-03-01 14:26:14.016 -06:00] [INF] [1] Main: Operating system: "Linux"
[2021-03-01 14:26:14.016 -06:00] [INF] [1] Main: Architecture: X64
[2021-03-01 14:26:14.017 -06:00] [INF] [1] Main: 64-Bit Process: True
[2021-03-01 14:26:14.017 -06:00] [INF] [1] Main: User Interactive: True
[2021-03-01 14:26:14.017 -06:00] [INF] [1] Main: Processor count: 16
[2021-03-01 14:26:14.017 -06:00] [INF] [1] Main: Program data path: "/var/lib/jellyfin"
[2021-03-01 14:26:14.017 -06:00] [INF] [1] Main: Web resources path: "/usr/share/jellyfin-web"
[2021-03-01 14:26:14.017 -06:00] [INF] [1] Main: Application directory: "/usr/lib64/jellyfin/"
[2021-03-01 14:26:14.115 -06:00] [INF] [1] Emby.Server.Implementations.AppBase.BaseConfigurationManager: Setting cache path: "/var/cache/jellyfin"
[2021-03-01 14:26:14.145 -06:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Defined LAN addresses : "[127.0.0.1/32,10.0.0.0/24,2600:6c56:7108:14b1::/64,::1/128,127.0.0.1/8]"
[2021-03-01 14:26:14.145 -06:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Defined LAN exclusions : "[]"
[2021-03-01 14:26:14.146 -06:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using LAN addresses: "[127.0.0.1/32,10.0.0.0/24,2600:6c56:7108:14b1::/64,::1/128,127.0.0.1/8]"
[2021-03-01 14:26:14.149 -06:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using bind addresses: "[]"
[2021-03-01 14:26:14.149 -06:00] [INF] [1] Jellyfin.Networking.Manager.NetworkManager: Using bind exclusions: "[]"
</pre>

**Comment by BaronGreenback** [comment]
Could you try disabling IPv4 and enabling IPv6?

I've checked the code and cannot find anything. I'm also running IP4 and IP6 here without issue.

**Comment by BaronGreenback** [comment]
@aussetg What platform are you running on?

**Comment by mooninite** [comment]
Disabling IPv4 and enabling IPv6 still lets Jellyfin start. Now I get these messages in the log:

<pre>
[2021-03-01 16:44:11.317 -06:00] [WRN] [17] Jellyfin.Networking.Manager.NetworkManager: IPv4 is disabled in Jellyfin, but enabled in the OS. This may affect how the interface is selected.
</pre>

Here's the listening socket when IPv6 is enabled:
<pre>
LISTEN     0      512                                            *:8096                                          *:*
</pre>

Here's the listening socket when IPv4 is enabled:
<pre>
LISTEN     0      512                                      0.0.0.0:8096                                    0.0.0.0:*
</pre>

**Comment by BaronGreenback** [comment]
Well that took some digging!

It looks like .NET's kestrel source code implements dualmode sockets, which only seem to error on non-window's system!

```
// Kestrel expects IPv6Any to bind to both IPv6 and IPv4
if (endPoint.Address == IPAddress.IPv6Any)
{
    listenSocket.DualMode = true;
}
```

Will work on a PR.

**Comment by BaronGreenback** [comment]
https://github.com/jellyfin/jellyfin/pull/5345

**Comment by mooninite** [comment]
Yes, that PR fixes it for me. Thank you for the quick fix. No other errors in the log. Successfully tested video playback with a HD Homerun live stream.

**Comment by BaronGreenback** [comment]
Just a note for others searching on this issue. 

A workaround would be to manually specifiy the interfaces to use in the settings.

**Comment by cvium** [comment]
Fixed by #5345
